File: lab/javascript/fatorial/grande.js
function multiplicaVetorPorInteiro(vetor, k) {
  let carry = 0;
  for (let i = 0; i < vetor.length; i++) {
    let produto = vetor[i] * k + carry;
    vetor[i] = produto % 10;
    carry = Math.floor(produto / 10);
  }
  // drena o carry
  while (carry > 0) {
    vetor.push(carry % 10);
    carry = Math.floor(carry / 10);
  }
}

function bigFatorial(n) {
  if (n === 0 || n === 1) return [1];
  const resultado = [1];

  for (let multiplicador = 2; multiplicador <= n; multiplicador++) {
    multiplicaVetorPorInteiro(resultado, multiplicador);
  }

  return resultado; // ainda invertido
}

// Exemplo: 10!
const fat = bigFatorial(10);
console.log(fat.reverse().join('')); // 3628800

File: lab/python/fatorial/grande.py
def multiplica_vetor_por_inteiro(vetor, k):
    carry = 0
    # percorre TODOS os dígitos (menos significativo -> mais significativo)
    for i in range(len(vetor)):
        produto = vetor[i] * k + carry
        vetor[i] = produto % 10
        carry = produto // 10
    # drena o carry restante
    while carry > 0:
        vetor.append(carry % 10)
        carry //= 10


def big_fatorial(n):
    if n == 0 or n == 1:
        return [1]  # vetor [1] representa 1
    resultado = [1]
    for multiplicador in range(2, n + 1):  # vai até n
        multiplica_vetor_por_inteiro(resultado, multiplicador)
    return resultado  # ainda invertido (LSB -> MSB)


fat = big_fatorial(10)
print(''.join(map(str, reversed(fat))))  # 3628800
